src/components/utils/src/timer.cc

```diff
@@ -44,175 +44,176 @@ CREATE_LOGGERPTR_GLOBAL(logger_, "Utils")
   LOG4CXX_AUTO_TRACE(logger_);
   LOG4CXX_DEBUG(logger_,
                 "Timer has finished counting. Timeout(ms): "
-                    << static_cast<uint32_t>(timeout_ms_));
-  {
-    // Task locked by own lock because from this task in callback we can
-    // call Stop of this timer and get DeadLock
-    sync_primitives::AutoLock auto_lock(task_lock_);
-    DCHECK(task_);
-    task_->run();
-  }
-  sync_primitives::AutoLock auto_lock(lock_);
-  if (is_running_) {
-    const bool stop_result = StopUnsafe();
-    DCHECK_OR_RETURN_VOID(stop_result);
-  }
-  if (repeatable_) {
-    StartUnsafe();
-  }
+                    << static_cast<uint32_t>(delegate_.get_timeout()));
+
+  DCHECK(task_.get());
+  task_->run();
 }
 
-void timer::Timer::SetTimeoutUnsafe(const timer::Milliseconds timeout) {
-  timeout_ms_ = (0u != timeout) ? timeout : 1u;
+timer::Milliseconds timer::Timer::GetTimeout() const {
+  return delegate_.get_timeout();
 }
 
-void timer::Timer::StartUnsafe() {
-  LOG4CXX_DEBUG(logger_, "Creating posix_timer in " << name_);
-  // Create new posix timer
-  timer_ = StartPosixTimer(*this, timeout_ms_);
-  DCHECK_OR_RETURN_VOID(timer_);
-  is_running_ = true;
+timer::Timer::TimerDelegate::TimerDelegate(Timer* timer)
+    : timer_(timer)
+    , timeout_milliseconds_(0)
+    , state_lock_(true)
+    , stop_flag_(false)
+    , restart_flag_(false)
+    , is_started_flag_(false)
+    , is_repeatable_(false) {}
+
+timer::Timer::TimerDelegate::~TimerDelegate() {
+  timer_ = NULL;
 }
 
-bool timer::Timer::StopUnsafe() {
-  LOG4CXX_DEBUG(logger_, "Stopping timer  " << name_);
-  //  Destroing of posix timer
-  if (StopPosixTimer(timer_)) {
-    is_running_ = false;
-    return true;
+void timer::Timer::TimerDelegate::threadMain() {
+  using sync_primitives::ConditionalVariable;
+  sync_primitives::AutoLock auto_lock(state_lock_);
+  stop_flag_ = false;
+  // Notify that we've already started
+  TimerDelegate::is_started_flag_ = true;
+  TimerDelegate::starting_condition_.NotifyOne();
+
+  while (!stop_flag_) {
+    // Sleep
+    int32_t wait_milliseconds_left = TimerDelegate::get_timeout();
+    LOG4CXX_DEBUG(logger_,
+                  "Milliseconds left to wait: " << wait_milliseconds_left);
+    ConditionalVariable::WaitStatus wait_status =
+        termination_condition_.WaitFor(auto_lock, wait_milliseconds_left);
+    // Quit sleeping or continue sleeping in case of spurious wake up
+    if (ConditionalVariable::kTimeout == wait_status ||
+        wait_milliseconds_left <= 0) {
+      LOG4CXX_DEBUG(logger_,
+                    "Timer has finished counting. Timeout(ms): "
+                        << wait_milliseconds_left);
+      DCHECK(timer_);
+      timer_->OnTimeout();
+    } else {
+      LOG4CXX_DEBUG(
+          logger_,
+          "Timeout reset force (ms): " << TimerDelegate::timeout_milliseconds_);
+    }
+
+    if (is_repeatable_) {
+      continue;
+    }
+    if (restart_flag_) {
+      restart_flag_ = false;
```

Comment:
@anosach-luxoft , to your class example need add: 
1. Oneshot and multi-shot timer functionality
2. Need add feature to restart one-shot timer with new timer.
3. Add functionality to stop and restart thread itself.
4. Add functionality to synchronously Start timer, otherwise current unit tests will fail.
5. Some service methods to set and get parameters.

So, if do it all, you will get my code.